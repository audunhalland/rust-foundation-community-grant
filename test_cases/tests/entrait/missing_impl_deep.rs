struct Impl<T>(T);

trait T1 {
    fn t1(&self);
}

fn f1(deps: &impl T2) {
    deps.t2();
}

impl<T> T1 for Impl<T>
where
    Self: T2,
{
    fn t1(&self) {
        f1(self)
    }
}

trait T2 {
    fn t2(&self);
}

fn f2(deps: &impl T3) {
    deps.t3();
}

impl<T> T2 for Impl<T>
where
    Self: T3,
{
    fn t2(&self) {
        f2(self)
    }
}

trait T3 {
    fn t3(&self);
}

impl<T> T3 for Impl<T>
where
    Self: T4,
{
    fn t3(&self) {
        f3(self)
    }
}

fn f3(deps: &impl T4) {}

trait T4 {}

fn main() {
    let app = Impl(());
    // Note: The reason this fails is that T4 is not implemented for Impl<T>:
    app.t1();
}
