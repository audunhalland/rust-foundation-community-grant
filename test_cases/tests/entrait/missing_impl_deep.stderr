error[E0599]: the method `t1` exists for struct `Impl<()>`, but its trait bounds were not satisfied
  --> tests/entrait/missing_impl_deep.rs:57:9
   |
1  | struct Impl<T>(T);
   | --------------
   | |      |
   | |      method `t1` not found for this struct
   | doesn't satisfy `Impl<()>: T1`
   | doesn't satisfy `Impl<()>: T2`
...
57 |     app.t1();
   |         ^^ method cannot be called on `Impl<()>` due to unsatisfied trait bounds
   |
note: trait bound `Impl<()>: T2` was not satisfied
  --> tests/entrait/missing_impl_deep.rs:13:11
   |
11 | impl<T> T1 for Impl<T>
   |         --     -------
12 | where
13 |     Self: T2,
   |           ^^ unsatisfied trait bound introduced here
note: the following trait must be implemented
  --> tests/entrait/missing_impl_deep.rs:20:1
   |
20 | trait T2 {
   | ^^^^^^^^
